### ISSUE TITLE
Typescript definition files sometimes not picked correctly

### ISSUE THREAD

**Issue by dbaeumer** [Microsoft/TypeScript#9539]
_From @csvn on July 5, 2016 17:28_

- VSCode Version: 1.2.1 (and 1.3.0-insider)
- OS Version: Windows 8.1 Pro

![typescript-library-errors](https://cloud.githubusercontent.com/assets/8770194/16593212/9cf4bf22-42e4-11e6-9acd-1b185cc2980c.gif)

Seemingly at random when I open a typescript file, `d.ts` files will not work correctly. `Reload typescript project`, `Reload window` or navigating to `tsconfig.json` and back may remove red squiggly lines. But when one file works, another usually do not. This means I almost always have red errors in vscode for imported files or even `lib.es6.d.ts` objects/methods.

The only easy place where I can reproduce it is in my repo (see .gif above).

**Steps to Reproduce:**

1. Clone https://github.com/csvn/ng-esm.git and run `npm install`
2. Open folder with vscode
3. Navigate between different `.ts` files under `src` folder
4. vscode should complain about: unknown namespaces, Symbol, Reflect etc
5. Reload window/typescript
6. The current file should be fine, but navigating to other `.ts` files produce similar errors

I have the same issue on two computers with both insider and regular vscode. I believe this to be an issue with vscode due to typescript building without issues, and due to the randomness of the errors.

_Copied from original issue: Microsoft/vscode#8764_

**Comment by dbaeumer** [comment]
Moving to TS team since TS smartness is powered by tsserver.

**Comment by JamieClayton** [comment]
I can replicated the intellisense warnings in Visual Studio 2015 Update 3 and TS 1.8.34 plugin on Windows 10. The TS compiles and works correctly in the web application, but the IDE shows incorrect warnings (red squiggles under references). 'Could not find symbol xzy in 'global' module.  The same solution on a colleagues workstation doesn't exhibit IDE warning.

**Comment by mhegazy** [comment]
@csvn, is there a reason the tsconfig.json does not include all your files? the files included in this tsconfig.json are only "main" and the files it imports. and as far as i can tell there are files that are *not* covered by this tsconfig.json in your project.

```ts
  "files": [
    "main.ts",
    "../typings/index.d.ts"
  ]
```

**Comment by csvn** [comment]
@mhegazy I believed that the files would be included due to them being imported from the "main" file. I believe I followed the documentation as best I could, but it mostly said to use either "files" or "exclude" properties.

Compiling with `tsc` did not produce errors, so I did not believe the configuration was wrong. Is it common to include all files? I also have a "example" folder with it's own tsconfig and typescript files, would they not clash if all files are included?

**Comment by mhegazy** [comment]
> Compiling with tsc did not produce errors

consider running `tsc --listFiles` to see all the files that the compiler picks up

> Is it common to include all files

do not see why that would be wrong. it all depends on what you want to do.

>  I also have a "example" folder with it's own tsconfig and typescript files, would they not clash if all files are included?

you can exclude this folder using `"exclude"` property.

**Comment by csvn** [comment]
Running `tsc -p src --listFiles` correctly yields the main file and all imports (listed at the bottom).

> do not see why that would be wrong. it all depends on what you want to do.

I wanted to specify the app as clearly as I could, which seemed logical for me to be the main file. I thought it odd for files I had not yet imported to be included in the project automatically. But if it's the best way to do it I'll change it.

Just seems that there is some inconsistency if it compiles without errors, but the editor warns you about it intermittently.

----------------------------------

```
C:\Dev\ng-esm>tsc -p src --listFiles
C:/Users/csv/AppData/Roaming/npm/node_modules/typescript/lib/lib.es6.d.ts
C:/Dev/ng-esm/src/common.ts
C:/Dev/ng-esm/src/case.ts
C:/Dev/ng-esm/src/ng.ts
C:/Dev/ng-esm/src/register/component.ts
C:/Dev/ng-esm/src/register/config.ts
C:/Dev/ng-esm/src/register/filter.ts
C:/Dev/ng-esm/src/register/module.ts
C:/Dev/ng-esm/src/register/resolve.ts
C:/Dev/ng-esm/src/register/run.ts
C:/Dev/ng-esm/src/register/service.ts
C:/Dev/ng-esm/src/register/state.ts
C:/Dev/ng-esm/src/register/index.ts
C:/Dev/ng-esm/src/options.ts
src/main.ts
typings/globals/angular-material/index.d.ts
typings/globals/angular-ui-router/index.d.ts
typings/globals/angular/index.d.ts
typings/globals/jquery/index.d.ts
typings/index.d.ts
```

**Comment by csvn** [comment]
It's possible to remove this error by doing the following:

1. Updating TypeScript version to `^2.0.0` in `package.json`
2. Setting `"typescript.tsdk": "./node_modules/typescript/lib"` in vscode `settings.json`

Not sure if its the version or by using the setting that fixes the issue.

This workaround is fine for now for me. You can close the issue unless you want to look into it.

**Comment by mhegazy** [comment]
Just a note for posterity. The reason you need to update the VSCode setting is that there are two pieces at play here; the command line compiler you use to build your project at build time, i.e. `tsc`, this comes from your npm version. then there is the version of the tools used in VSCode. by default VSCode today comes with 1.8.10 (last 1.8 version). to get the same experience in both, you need to update the npm version and the `typescript.tssdk` in your settings.

this also applies to other IDE's see https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/Nightly%20Builds.md#updating-your-ide-to-use-the-nightly-builds for more details.

**Comment by RyanCavanaugh** [comment]
Is there anything left to do here? Everything OK?

**Comment by csvn** [comment]
@RyanCavanaugh Sorry, forgot about this thread. It may have been the different versions of `tsc` used by the editor and the build process that made it behave oddly. I think this issue is safe to close, as the changes above made the issue go away.

It's probably a good idea to have VSCode use the same typescript version anyway.

**Comment by RyanCavanaugh** [comment]
Thanks!
